### ISSUE TITLE
Implement complex types (value objects)

### ISSUE THREAD

**Issue by eveneveneven** [dotnet/efcore#13947]
Greetings, this is a question/feature request:

Per the Limitations section of the Value Conversions documentation: "There is currently no way to spread a conversion of one property to multiple columns or vice-versa."

Is something like this planned, or is there an alternative better way to do this?
I'm looking for a way to keep the backing fields out of my entity classes. And to not use strings in the configuration or public backing fields in my entity classes.

Here's a quick example from my current code working with a legacy DB:

From an entity class:

```c#
      private DateTime _lastChanged { get; set; }
      private int _lastChangedHour { get; set; }
      private int _lastChangedMinute { get; set; }      
      public DateTime LastChanged
      {
         get
         {
            if (_lastChanged.Hour != 0 || 
                _lastChanged.Minute != 0 || 
                _lastChanged.Second != 0)
               return _lastChanged;

            var returnDateTime = 
               new DateTime(
                  _lastChanged.Year, 
                  _lastChanged.Month, 
                  _lastChanged.Day, 
                  _lastChangedHour, 
                  _lastChangedMinute, 
                  0);
            return returnDateTime;
         }

         set
         {
            _lastChangedHour = value.Hour;
            _lastChangedMinute = value.Minute;
            _lastChanged = value;
         }
      }
```

From the entity's configuration:

```c#
            builder.Ignore(o => o.LastChanged);

            builder
               .Property<DateTime>("_lastChanged")
               .HasColumnName("SIST_ENDRET")
               .HasColumnType("datetime")
               .HasDefaultValue((DateTime) SqlDateTime.MinValue);

            builder
               .Property<int>("_lastChangedHour")
               .HasColumnName("SIST_ENDRET_TIM")
               .HasColumnType("decimal(2,0)")
               .HasConversion<decimal>();

            builder
               .Property<int>("_lastChangedMinute")
               .HasColumnName("SIST_ENDRET_MIN")
               .HasColumnType("decimal(2,0)")
               .HasConversion<decimal>();
```

Thanks for your time!

**Comment by ajcvickers** [comment]
@divega to look for dupe.

**Comment by divega** [comment]
This issue could be considered a duplicate of #9906 "Use C# structs or classes as value objects", based on the triage notes at https://github.com/aspnet/EntityFrameworkCore/issues/9906#issuecomment-334278233, and the fact that #13067 was previously closed in favor of #9906. 

But I personally don't think the association is clear, especially given that we don't have a design for #9906. I would prefer to keep this issue open to represent the desire to have value conversions that spread multiple columns, so that customers can vote for it and discuss it independently of structs and value objects.

**Comment by mcd8604** [comment]
Having value conversions spread over multiple columns is exactly what I was hoping I could do. I have a property type (Length) in my data model which is defined as a struct in a third party library. It contains two critical data fields - the value (double) and the unit (enum). I simply want to map the value to a double column and the string-converted-enum to another column. How could I map this property without being able to create an entity or mapping to multiple columns?

**Comment by ajcvickers** [comment]
Notes from planning: we decided to close this in favor of #9906 (Use C# structs or classes as value objects) or vice versa, but reading again the comment from @divega above, I think it does make sense to keep this as separate issues. However, we'll likely work on them together.

**Comment by NichUK** [comment]
+1 please...
I have a custom value type for some finance work, which I'd like to be able to store in a readable fashion, but right now my only option seems to be storing as a byte array. It works, but it's not what I'd like. Flattening my type into 3 columns in the main table would be perfect.

**Comment by 1234Georg** [comment]
An additional note: For me it would be very interesting to map a combined key (2 columns) with a conversion in a single class or struct. That also means, that I would need to query this combined key, and the restriction should be evaluated in the database and not on the client. Otherwise it would load way to much data. An example is in the issue #15962.

**Comment by GregOCC** [comment]
Another +1 for this. (Sorry I can't find a way to record that other than by commenting!)

**Comment by spaasis** [comment]
This would enable really painless audit trails for EF-Core, as what is currently standing in the way is that there are issues in getting the changed values of Value Objects. Related SO: https://stackoverflow.com/questions/60946496/ef-core-how-to-audit-trail-with-value-objects

**Comment by AndriySvyryd** [comment]
**Note to implementor:** Consider allowing AKs, FKs and indexes to use only some of the columns and mixing with other properties, see #11336

**Comment by vslee** [comment]
The following article describes a strategy which might improve the performance in the special case where SQL Server is used, and one of the columns is a date column:
https://sqlperformance.com/2018/03/sql-optimizer/dealing-with-date-and-time

**Comment by mguinness** [comment]
Another area where this could be useful is storing [DateTimeOffset](https://mysqlconnector.net/troubleshooting/datetime-storage/#datetimeoffset-notes) into two fields for databases like MySQL which have no equivalent data type.

**Comment by roji** [comment]
@Joren-Thijs-KasparSolutions this issue is in the 7.0.0 milestone, meaning that it's currently in the plan for the upcoming version of EF Core.

**Comment by ajcvickers** [comment]
It is with much sadness that we are cutting this feature from the EF7 release. We know people are waiting on this, and it will be a high priority for EF8.

**Comment by jscarle** [comment]
@AndriySvyryd Will this make the cut for preview 6?

**Comment by AndriySvyryd** [comment]
@jscarle No. It's still unclear in which preview it will be useable.

**Comment by DumboJet** [comment]
So, is this implemented now?
I don't see the docs [here](https://learn.microsoft.com/en-us/ef/core/modeling/value-conversions?tabs=data-annotations#composite-value-objects) updated.
How is this done?

**Comment by ajcvickers** [comment]
@DumboJet Docs are behind. See [Value objects using Complex Types](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-8.0/whatsnew#value-objects-using-complex-types).
